### PR TITLE
Fixed bugs for new Contact Information Service

### PR DESCRIPTION
**Pull request by RachalCassity**
**Note**: Delete the description statements, complete each step. **None are optional**, but can be justified as to why they cannot be completed as written. Provide known gaps to testing that may raise the risk of merging to production.

## Summary

CIV2 = VAProfile Contact Information V2

50% of files in this PR are VCR cassette modifications. Only 22 non-spec files have modifications.

- *This work is behind a feature toggle (flipper): YES

Internal testing was done for the new Contact Information Service. All bugs were addressed in this pull request.
Most of the spec files were updated to remove vet360_id == '1781151'. Using the defined user vet360_id in the user factory will work.

**CIV2 Service Modifications**
Only ICNS can be used for CIV2 requests, `icn_with_aaid`.

**VAProfile Policy Modifications**
Verified user, aka user with icn, has to be present to make CIV2 requests.

**VAProfile Model Modifications**
VAProfile::Models::Email
VAProfile::Models::Telephone
VAProfile::Models::V3::Address

The method, in_json_v2, was added to the VAProfile models listed above. The in_json_v2 method removes the vet360_id from VAProfile Contact Information V2 API requests. vet360_id is no longer necessary. All vet360_id params were removed from CIv2 requests. This will reduce any confusion regarding what is required for CIV2 requests.

**Spec Modifications**
Spec files were modified to use an icn for CIV2 requests. Any reference defining user.vet360_id was removed to avoid confusion in the future because vet360_id is no longer valid for CIV2 requests. All CIV2 specs use the user: :loa factory settings. :) 

**VCR Modifications**
VAPROFILE_URL was removed from the [VCR configuration](https://github.com/department-of-veterans-affairs/vets-api/blob/e06ada225606e0be76495c4dd22398b8635bd3de/spec/support/vcr.rb#L38). The same url is defined for VAPROFILE_URL and VETS360_URL. All VCR VAPROFILE_URL cassettes were updated to use VETS360_URL.

## Related issue(s)

-  department-of-veterans-affairs/va.gov-team/issues/91488)

## Testing done

- [X] *New code is covered by unit tests*
- Profile information was updated to address new attributes with the new service

## Screenshots
_Note: Optional_

## What areas of the site does it impact?
*(Describe what parts of the site are impacted and*if*code touched other areas)*

## Acceptance criteria

- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [x]  Events are being sent to the appropriate logging solution
- [ ]  Documentation has been updated (link to documentation)
- [x]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

## Requested Feedback

(OPTIONAL)_What should the reviewers know in addition to the above. Is there anything specific you wish the reviewer to assist with. Do you have any concerns with this PR, why?_

### CODE REVIEW

File: lib/va_profile/v2/person/service.rb [hunk 1]
```diff
@@ -25,24 +25,40 @@ class Service < VAProfile::Service
...
+            raise 'User does not have a valid ICN with an Assigning Authority ID'
+          end
+        end
+
         private
 
         # @see https://ruby-doc.org/stdlib-2.3.0/libdoc/erb/rdoc/ERB/Util.html
         #
         def uuid_with_aaid
+          return "#{vet360_id}^PI^200VETS^USDVA" if vet360_id
```
Comment (RachalCassity) [inline lib/va_profile/v2/person/service.rb:58]: Use vet360_id or icn before uuids.

File: app/controllers/concerns/vet360/writeable.rb [hunk 2]
```diff
@@ -21,13 +21,10 @@ module Writeable
     #
     def write_to_vet360_and_render_transaction!(type, params, http_verb: 'post')
       output_rails_logs = Flipper.enabled?(:va_v3_contact_information_service, @current_user)
-      Rails.logger.info('Building vaprofile record') if output_rails_logs
```
Comment (RachalCassity) [inline app/controllers/concerns/vet360/writeable.rb:24]: Removed extra logs

File: app/models/async_transaction/va_profile/base.rb [hunk 3]
```diff
@@ -43,8 +43,12 @@ def self.last_requested_for_user(user)
       #
       def self.start(user, response)
         # vet360_id is no longer required for Contact Information API V2
-        source_id = Flipper.enabled?(:va_v3_contact_information_service, user) ? user.vet360_id || 0 : user.vet360_id
```
Comment (RachalCassity) [inline app/models/async_transaction/va_profile/base.rb:46]: Update source_id if using new service that does not require user to have a vet360_id

File: app/models/user.rb [hunk 4]
```diff
@@ -407,7 +407,7 @@ def va_profile_email
   end
 
   def vaprofile_contact_info
-    @vet360_contact_info ||= VAProfileRedis::V2::ContactInformation.for_user(self)
+    @vaprofile_contact_info ||= VAProfileRedis::V2::ContactInformation.for_user(self)
```
Comment (RachalCassity) [inline app/models/user.rb:410]: Necessary change.

File: app/policies/va_profile_policy.rb [hunk 5]
```diff
@@ -8,7 +8,7 @@ def access?
   def access_to_v2?
     if Flipper.enabled?(:va_v3_contact_information_service, user)
       # user vet360_id is no longer needed for Contact Information Api V2
-      user.vet360_id.present? || user.idme_uuid.present? || user.logingov_uuid.present?
+      user.vet360_id.present? || user.icn.present? || user.idme_uuid.present? || user.logingov_uuid.present?
```
Comment (RachalCassity) [inline app/policies/va_profile_policy.rb:11]: This allows users to find info with any uuid

File: app/services/users/profile.rb [hunk 6]
```diff
@@ -131,7 +133,7 @@ def vet360_contact_information
         work_phone: person.work_phone,
         temporary_phone: person.temporary_phone,
         fax_number: person.fax_number,
-        text_permission: person.text_permission
+        text_permission: contact_info_v2_enabled ? nil : person.text_permission
```
Comment (RachalCassity) [inline app/services/users/profile.rb:136]: Text permission was removed from CI API V2

File: lib/va_profile/models/email.rb [hunk 7]
```diff
@@ -22,7 +22,7 @@ class Email < Base
       attribute :transaction_id, String
       attribute :updated_at, Common::ISO8601Time
       attribute :vet360_id, String
-      attribute :vaprofile_id, String
+      attribute :va_profile_id, String
```
Comment (RachalCassity) [inline lib/va_profile/models/email.rb:25]: Fixed attribute

File: lib/va_profile/models/email.rb [hunk 8]
```diff
@@ -42,7 +42,8 @@ def in_json
             originatingSourceSystem: SOURCE_SYSTEM,
             sourceSystemUser: @source_system_user,
             sourceDate: @source_date,
-            vet360Id: @vet360_id,
+            vet360Id: @vet360_id || @vaProfileId,
```
Comment (RachalCassity) [inline lib/va_profile/models/email.rb:45]: Needed to ensure all data is transferred to both services.

File: lib/va_profile/models/email.rb [hunk 9]
```diff
@@ -62,7 +63,8 @@ def self.build_from(body)
           source_date: body['source_date'],
           transaction_id: body['tx_audit_id'],
           updated_at: body['update_date'],
-          vet360_id: body['vet360_id']
+          vet360_id: body['vet360_id'] || body['va_profile_id'],
```
Comment (RachalCassity) [inline lib/va_profile/models/email.rb:66]: Needed to ensure all data is transferred to both services.

File: lib/va_profile/models/telephone.rb [hunk 10]
```diff
@@ -112,7 +113,9 @@ def in_json
             textMessageCapableInd: @is_textable,
             textMessagePermInd: @is_text_permitted,
             ttyInd: @is_tty,
-            vet360Id: @vet360_id,
+            vet360Id: @vet360_id || @vaProfileId,
```
Comment (RachalCassity) [inline lib/va_profile/models/telephone.rb:116]: Needed to ensure all data is transferred to both services.

File: lib/va_profile/models/telephone.rb [hunk 11]
```diff
@@ -141,7 +144,8 @@ def self.build_from(body)
           transaction_id: body['tx_audit_id'],
           is_tty: body['tty_ind'],
           updated_at: body['update_date'],
-          vet360_id: body['vet360_id'],
+          vet360_id: body['vet360_id'] || body['va_profile_id'],
```
Comment (RachalCassity) [inline lib/va_profile/models/telephone.rb:147]: Needed to ensure all data is transferred to both services.

File: lib/va_profile/models/v3/address.rb [hunk 12]
```diff
@@ -38,7 +38,8 @@ def in_json
             originatingSourceSystem: SOURCE_SYSTEM,
             sourceSystemUser: @source_system_user,
             sourceDate: @source_date,
-            vet360Id: @vet360_id,
+            vet360Id: @vet360_id || @vaProfileId,
```
Comment (RachalCassity) [inline lib/va_profile/models/v3/address.rb:41]: Needed to ensure all data is transferred to both services.

File: lib/va_profile/models/v3/address.rb [hunk 13]
```diff
@@ -88,7 +89,8 @@ def self.build_from(body)
             state_code: body['state_code'],
             transaction_id: body['tx_audit_id'],
             updated_at: body['update_date'],
-            vet360_id: body['vet360_id'],
+            vet360_id: body['vet360_id'] || body['va_profile_id'],
```
Comment (RachalCassity) [inline lib/va_profile/models/v3/address.rb:92]: Needed to ensure all data is transferred to both services.

File: lib/va_profile/person/service.rb [hunk 14]
```diff
@@ -3,9 +3,7 @@
 require 'common/client/base'
 require 'common/client/concerns/monitoring'
 require 'va_profile/contact_information/configuration'
-require 'va_profile/v2/contact_information/configuration'
```
Comment (RachalCassity) [inline lib/va_profile/person/service.rb:6]: New service requires v3/person/service

File: lib/va_profile/v2/person/service.rb [hunk 15]
```diff
@@ -25,24 +25,40 @@ class Service < VAProfile::Service
         # @return [VAProfile::V2::ContactInformation::PersonTransactionResponse]
         # response wrapper around a transaction object
         #
-        def init_vet360_id
+        def init_vet360_id(icn = nil)
```
Comment (RachalCassity) [inline lib/va_profile/v2/person/service.rb:28]: Fixed person service to find by icn

File: lib/va_profile/v2/person/service.rb [hunk 16]
```diff
@@ -25,24 +25,40 @@ class Service < VAProfile::Service
...
-            raw_response = perform(:post, "#{MPI::Constants::VA_ROOT_OID}/#{ERB::Util.url_encode(uuid_with_aaid)}",
-                                   empty_body)
+            raw_response = perform(:post, encode_url!(icn), empty_body)
             VAProfile::V2::ContactInformation::PersonTransactionResponse.from(raw_response, @user)
           end
         rescue => e
           handle_error(e)
         end
 
+        def encode_url!(icn)
```
Comment (RachalCassity) [inline lib/va_profile/v2/person/service.rb:37]: This all was copied from the previous person service and is needed for now.

File: app/services/users/profile.rb (hunk 6 above)
Comment (LindseySaari) [reply app/services/users/profile.rb:136]: Is there anything equivalent for this?

**Review by LindseySaari — DISMISSED** [review]

File: app/services/users/profile.rb (hunk 6 above)
Comment (RachalCassity) [reply app/services/users/profile.rb:136]: Nope, it was removed awhile back. The text permission service is already enabled.

Comment (bosawt) [reply app/services/users/profile.rb:136]: Just remove this line. I looked on the frontend, there's nothing using `text_permission` or `textPermission` except tests

Comment (bosawt) [reply app/services/users/profile.rb:136]: Actually I'm going to create tickets for identity backend and frontend to remove a ton of these fields from the `/user` serializer and from the frontend tests. I don't see usage of basically any of these fields on the frontend

Comment (bosawt) [reply app/services/users/profile.rb:136]: @RachalCassity Basically do what you want with this field, I'm gonna get my team to remove it from profile soon.

Comment (RachalCassity) [reply app/services/users/profile.rb:136]: @bosawt I will remove this line and make sure it doesn't break the current Contact Information Service.

Comment (RachalCassity) [reply app/services/users/profile.rb]: @bosawt Done! Testing was completed locally when the va_v3_contact_information_service flipper was turned off and on.

File: app/models/async_transaction/va_profile/base.rb [hunk 17]
```diff
@@ -43,8 +43,12 @@ def self.last_requested_for_user(user)
...
-
+        source_id = if Flipper.enabled?(:va_v3_contact_information_service,
+                                        user)
+                      user.vet360_id || user.uuid
```
Comment (RachalCassity) [inline app/models/async_transaction/va_profile/base.rb:48]: vet360_id is not required for the CI V2 service

File: app/policies/va_profile_policy.rb (hunk 5 above)
Comment (bosawt) [reply app/policies/va_profile_policy.rb:11]: I was looking at the original requirements to contacting the VA Profile service. It used to require a verified user, is that still the case?

If so, we probably shouldn't have an OR for idme_uuid/logingov_uuid, as that allows access to every single user on va.gov, not necessarily only verified users

Comment (bosawt) [reply app/policies/va_profile_policy.rb:11]: Additionally, every verified user will have ICN. It seems like things could be simplified quite a bit if the service was only ever called with ICN.

**Review by tpharrison — DISMISSED** [review]
LGTM 🚀

**Review by jperk51 — DISMISSED** [review]

File: lib/va_profile/v2/contact_information/service.rb [hunk 18]
```diff
@@ -212,11 +212,8 @@ def get_person_transaction_status(transaction_id)
 
         private
 
-        def uuid_with_aaid
-          return "#{@user.idme_uuid}^PN^200VIDM^USDVA" if @user.idme_uuid
-          return "#{@user.logingov_uuid}^PN^200VLGN^USDVA" if @user.logingov_uuid
-
-          "#{vet360_id}^PI^200VETS^USDVA"
+        def icn_with_aaid
```
Comment (RachalCassity) [inline lib/va_profile/v2/contact_information/service.rb:215]: Require ICN for CIV2

File: app/services/users/profile.rb [hunk 19]
```diff
@@ -118,20 +118,21 @@ def form_526_required_identifiers
...
+        va_profile_id: user.vet360_id,
         email: person.email,
         residential_address: person.residential_address,
         mailing_address: person.mailing_address,
         mobile_phone: person.mobile_phone,
         home_phone: person.home_phone,
         work_phone: person.work_phone,
         temporary_phone: person.temporary_phone,
-        fax_number: person.fax_number,
-        text_permission: person.text_permission
```
Comment (RachalCassity) [inline app/services/users/profile.rb:134]: Text permissions is no longer used.

File: spec/lib/va_profile/v2/contact_information/service_spec.rb [hunk 20]
```diff
@@ -56,44 +54,8 @@
     end
   end
 
-  describe '#get_person error' do
```
Comment (RachalCassity) [inline spec/lib/va_profile/v2/contact_information/service_spec.rb:59]: Moved to bottom of this spec file

File: spec/requests/v0/profile/telephones_spec.rb [hunk 21]
```diff
@@ -480,18 +473,16 @@
...
     before do
       Flipper.enable(:va_v3_contact_information_service)
-      telephone.id = id_in_cassette
     end
 
     after do
       Flipper.disable(:va_v3_contact_information_service)
     end
 
-    let(:user) { build(:user, :loa3, icn: '64762895576664260') }
```
Comment (RachalCassity) [inline spec/requests/v0/profile/telephones_spec.rb:490]: Not sure why a different icn was used. This was removed to use the icn defined in the :loa user factory

**Review by tpharrison — APPROVED** [review]

**Pull request merged**